src/Canvas.cc

```diff
@@ -227,132 +223,110 @@ Canvas::ToPngBufferAsync(uv_work_t *req) {
   write_to_jpeg_buffer(closure->canvas->surface(), closure);
 }
 #endif
 
-/*
- * EIO after toBuffer callback.
- */
+static void
+parsePNGArgs(Napi::Value arg, PngClosure& pngargs) {
+  if (arg.IsObject()) {
+    Napi::Object obj = arg.As<Napi::Object>();
+    Napi::Value cLevel;
 
-void
-Canvas::ToBufferAsyncAfter(uv_work_t *req) {
-  Nan::HandleScope scope;
-  Nan::AsyncResource async("canvas:ToBufferAsyncAfter");
-  Closure* closure = static_cast<Closure*>(req->data);
-  delete req;
-
-  if (closure->status) {
-    Local<Value> argv[1] = { Canvas::Error(closure->status) };
-    closure->cb.Call(1, argv, &async);
-  } else {
-    Local<Object> buf = Nan::CopyBuffer((char*)&closure->vec[0], closure->vec.size()).ToLocalChecked();
-    Local<Value> argv[2] = { Nan::Null(), buf };
-    closure->cb.Call(sizeof argv / sizeof *argv, argv, &async);
-  }
-
-  closure->canvas->Unref();
-  delete closure;
-}
-
-static void parsePNGArgs(Local<Value> arg, PngClosure& pngargs) {
-  if (arg->IsObject()) {
-    Local<Object> obj = Nan::To<Object>(arg).ToLocalChecked();
-
-    Local<Value> cLevel = Nan::Get(obj, Nan::New("compressionLevel").ToLocalChecked()).ToLocalChecked();
-    if (cLevel->IsUint32()) {
-      uint32_t val = Nan::To<uint32_t>(cLevel).FromMaybe(0);
+    if (obj.Get("compressionLevel").UnwrapTo(&cLevel) && cLevel.IsNumber()) {
+      uint32_t val = cLevel.As<Napi::Number>().Uint32Value();
       // See quote below from spec section 4.12.5.5.
       if (val <= 9) pngargs.compressionLevel = val;
     }
 
-    Local<Value> rez = Nan::Get(obj, Nan::New("resolution").ToLocalChecked()).ToLocalChecked();
-    if (rez->IsUint32()) {
-      uint32_t val = Nan::To<uint32_t>(rez).FromMaybe(0);
+    Napi::Value rez;
+    if (obj.Get("resolution").UnwrapTo(&rez) && rez.IsNumber()) {
+      uint32_t val = rez.As<Napi::Number>().Uint32Value();
       if (val > 0) pngargs.resolution = val;
     }
 
-    Local<Value> filters = Nan::Get(obj, Nan::New("filters").ToLocalChecked()).ToLocalChecked();
-    if (filters->IsUint32()) pngargs.filters = Nan::To<uint32_t>(filters).FromMaybe(0);
+    Napi::Value filters;
+    if (obj.Get("filters").UnwrapTo(&filters) && filters.IsNumber()) {
+      pngargs.filters = filters.As<Napi::Number>().Uint32Value();
+    }
 
-    Local<Value> palette = Nan::Get(obj, Nan::New("palette").ToLocalChecked()).ToLocalChecked();
-    if (palette->IsUint8ClampedArray()) {
-      Local<Uint8ClampedArray> palette_ta = palette.As<Uint8ClampedArray>();
-      pngargs.nPaletteColors = palette_ta->Length();
+    Napi::Value palette;
+    if (obj.Get("palette").UnwrapTo(&palette) && palette.IsTypedArray()) {
+      Napi::TypedArray palette_ta = palette.As<Napi::TypedArray>();
+      pngargs.nPaletteColors = palette_ta.ElementLength();
       if (pngargs.nPaletteColors % 4 != 0) {
         throw "Palette length must be a multiple of 4.";
       }
+      pngargs.palette = static_cast<uint8_t*>(palette_ta.ArrayBuffer().Data());
```

Comment:
```suggestion
      pngargs.palette = palette_ta.Data();
```

(the typedarray's offset might not be 0)